commit 64082ca923f8
Author: Andreas Pehrson <pehrsons@mozilla.com>
Date:   Mon May 7 11:33:07 2018 +0200

    Bug 1458166 - Clear out tracks while destroying MediaStream. r=bryce, a=RyanVM
    
    MediaStream::Destroy() is part of a controlled shutdown sequence.
    If there are still tracks with content beyond Destroy() they will
    only get caught by the dtor, which may be on CC shutdown and too late.
    
    MozReview-Commit-ID: GV6XRiTCIRk
    
    --HG--
    extra : source : da84235630724b32d2e619c4ae6b7ae34a4c5c2d
---
 dom/media/MediaStreamGraph.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git dom/media/MediaStreamGraph.cpp dom/media/MediaStreamGraph.cpp
index 1c6683c966084..9cdb50e8abc98 100644
--- dom/media/MediaStreamGraph.cpp
+++ dom/media/MediaStreamGraph.cpp
@@ -2054,6 +2054,7 @@ MediaStream::DestroyImpl()
   for (int32_t i = mConsumers.Length() - 1; i >= 0; --i) {
     mConsumers[i]->Disconnect();
   }
+  mTracks.Clear();
   mGraph = nullptr;
 }
 
commit d18fab7f3fa9
Author: Andreas Pehrson <pehrsons@mozilla.com>
Date:   Mon May 7 11:36:35 2018 +0200

    Bug 1458166 - Remove listeners before destroying MediaStreams. r=bryce, a=RyanVM
    
    If listeners are still registered to a MediaStream on MediaStream::Destroy
    (triggered by MediaStream::UnregisterUser below), they will catch and act
    on further events from the MediaStream (such as a track ending).
    
    This may dispatch runnables that are unnecessary since we know we are
    shutting down.
    
    If we first remove the listeners from the MediaStream we will never see
    said events.
    
    MozReview-Commit-ID: IZ1kENqL2C8
    
    --HG--
    extra : source : 8dc7d2696a29c0752066ebe67962b1ad323c4153
---
 dom/media/DOMMediaStream.cpp | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git dom/media/DOMMediaStream.cpp dom/media/DOMMediaStream.cpp
index f69324d416a8f..febb527556701 100644
--- dom/media/DOMMediaStream.cpp
+++ dom/media/DOMMediaStream.cpp
@@ -455,10 +455,16 @@ DOMMediaStream::Destroy()
 {
   LOG(LogLevel::Debug, ("DOMMediaStream %p Being destroyed.", this));
   if (mOwnedListener) {
+    if (mOwnedStream) {
+      mOwnedStream->RemoveListener(mOwnedListener);
+    }
     mOwnedListener->Forget();
     mOwnedListener = nullptr;
   }
   if (mPlaybackListener) {
+    if (mPlaybackStream) {
+      mPlaybackStream->RemoveListener(mPlaybackListener);
+    }
     mPlaybackListener->Forget();
     mPlaybackListener = nullptr;
   }
